hangupsbot/plugins/botalive.py

```diff
@@ -1,142 +1,171 @@
-import asyncio, datetime, logging, time
+"""plugin to watermark conversations periodically determined by config entry"""
 
+import asyncio
+import datetime
 import hangups
-
+import logging
 import plugins
-
-import threadmanager
+import random
 
 logger = logging.getLogger(__name__)
 
-
-internal = {}
-
-
 def _initialise(bot):
     config_botalive = bot.get_config_option("botalive") or {}
     if not config_botalive:
         return
 
-    _new_config = {}
+    if not bot.memory.exists(["conv_data"]):
+        # should not come to this, but create it once as we need to store data for each conv in it
+        bot.memory.set_by_path(["conv_data"], {})
+        bot.memory.save()
+
+    # backwards compatibility
     if isinstance(config_botalive, list):
+        _new_config = {}
         if "admins" in config_botalive:
             _new_config["admins"] = 900
         if "groups" in config_botalive:
             _new_config["groups"] = 10800
+        bot.config.set_by_path(["botalive"], _new_config)
+        bot.config.save()
         config_botalive = _new_config
 
-    if "admin" in config_botalive and config_botalive["admin"] < 60:
-        config_botalive["admin"] = 60
-    if "groups" in config_botalive and config_botalive["groups"] < 60:
-        config_botalive["groups"] = 60
+    if "admins" not in config_botalive and "groups" not in config_botalive:
+        return
+
+    watermarkUpdater = watermark_updater(bot)
+
+    if "admins" in config_botalive:
+        if config_botalive["admins"] < 60:
+            config_botalive["admins"] = 60
+        plugins.start_asyncio_task(_periodic_watermark_queue_admins, watermarkUpdater, config_botalive["admins"])
+
+    if "groups" in config_botalive:
+        if config_botalive["groups"] < 60:
+            config_botalive["groups"] = 60
+        plugins.start_asyncio_task(_periodic_watermark_queue_groups, watermarkUpdater, config_botalive["groups"])
 
     logger.info("timing {}".format(config_botalive))
 
-    plugins.start_asyncio_task(_periodic_watermark_update, config_botalive)
+    watch_event_types = [
+        "message",
+        "membership",
+        "rename"
+        ]
+    for event_type in watch_event_types:
+        plugins.register_handler(_log_message, event_type)
 
+def _log_message(bot, event, command):
+    """log time to conv_data of event conv"""
 
-@asyncio.coroutine
-def _periodic_watermark_update(bot, config_botalive):
-    """runs in a separate thread - to prevent the processor from being
-    consumed entirely, we sleep for 5 seconds on each loop iteration"""
+    conv_id = str(event.conv_id)
+    if not bot.memory.exists(["conv_data", conv_id]):
+        bot.memory.set_by_path(["conv_data", conv_id], {})
+        bot.memory.save()
+    bot.memory.set_by_path(["conv_data", conv_id, "botalive"], datetime.datetime.now().timestamp())
+    # not worth a dump to disk, skip bot.memory.save()
```

Comment:
I'm still reluctant to use memory.json because i feel that it's not justified at the moment - to me, a full one-off update on bot restart is not a bug or deficiency, it's a feature. moreover, I don't see  a scenario where the bot will "forget" to update something, since it doesn't use actually use memory.json or an in-memory store to keep track of which conversations to watermark. Perhaps if you elaborated further how this were to happen as I may have missed something when evaluating this code.